Introduce a simple request evaluator.

DougGregor opened the pull request:
Meant as a replacement for the barely-started iterative type checker,
introduce a simpler "evaluator" that can evaluate individual requests
(essentially, function objects with some additional API), caching
results as appropriate and detecting cycles. Start stubbing out what
is involved in replacing the iterative type checker with this infrastructure.

DougGregor committed:
[Evaluator] Introduce a simple request evaluator.

Meant as a replacement for the barely-started iterative type checker,
introduce a simpler "evaluator" that can evaluate individual requests
(essentially, function objects with some additional API), caching
results as appropriate and detecting cycles.

huonw commented on include/swift/Basic/AnyValue.h:
```diff
@@ -0,0 +1,118 @@
+namespace swift {
+
+/// Stores a value of any type satisfies a small set of requirements
+/// (currenrly, just equatability).
```
currenrly -> currently.

huonw commented on include/swift/Basic/AnyValue.h:
```diff
@@ -0,0 +1,118 @@
+
+  /// Cast to a specific (known) type.
+  template<typename T>
+  const T &castTo() const {
```
This could have an assert on the `TypeId<T>::value`.

graydon commented on include/swift/AST/AnyRequest.h:
```diff
@@ -0,0 +1,227 @@
+///   - Hashing support (hash_value)
+///   - TypeID support (see swift/Basic/TypeID.h)
+///   - Cycle diagnostics operations:
+///
```
Did you forget 'OutputType breakCycle() const' here? It looks like it's present in a similar comment down below.

DougGregor replied:
`AnyRequest` itself doesn't need it, because we only call the `breakCycle` operation on a request before it's been type-erased. We can add it here if we need it later.

graydon commented on include/swift/AST/Evaluator.h:
```diff
@@ -0,0 +1,264 @@
+///
+///       void diagnoseCycle(DiagnosticEngine &diags) const;
+///       void noteCycleStep(DiagnosticEngine &diags) const;
+///       OutputType breakCycle() Const;
```
s/Const/const/

graydon commented on include/swift/AST/Evaluator.h:
```diff
@@ -0,0 +1,264 @@
+///     static const bool isEverCached;
+///
+///       When false, the request's result will never be cached. Note that this
+///       also precludes cycle detection.  When true, the result will be
```
Why does not-caching preclude cycle detection?

DougGregor replied:
Ah, it doesn't have to. I'm using the cache to detect the presence of a cycle, but there's no reason I have to do that... I have the stack of results to be satisfied and can unique there. Thanks!

huonw commented on include/swift/Basic/TypeID.h:
```diff
@@ -0,0 +1,99 @@
+                                                                        \
+public:                                                                 \
+  static const uint64_t value =                                         \
+    (templateID << (depth * 8)) | TypeID<Arg1>::value;                  \
```
Shouldn't this be `depth * 16`. Also, could we could possibly avoid `depth` all together using `(TypeID<Arg1>::value << 16) | templateID`, assuming `value` is guaranteed to be non-zero (and/or `static_assert` that it's non-zero)?

DougGregor replied:
Good idea, I've done this.

huonw commented on include/swift/Basic/TypeID.h:
```diff
@@ -0,0 +1,99 @@
+
+/// Form a type ID given a zone and type value.
+constexpr uint8_t formTypeID(uint8_t zone, uint8_t type) {
+  return (zone << 8) | type;
```
Hm, this looks like overflow; should this return `uint16_t`? (and are casts needed to stop the intermediate overflowing?)

graydon commented on include/swift/AST/Evaluator.h:
```diff
@@ -0,0 +1,264 @@
+  /// request.
+  void diagnoseCycle(const AnyRequest &request);
+
+  /// Retrieve the result of produced by evaluating a request that can
```
s/result of/result/

graydon commented on include/swift/AST/Evaluator.h:
```diff
@@ -0,0 +1,264 @@
+    return getResultUncached(request);
+  }
+
+  /// Retrieve the result of produced by evaluating a request that
```
s/result of/result/

graydon commented on include/swift/AST/Evaluator.h:
```diff
@@ -0,0 +1,264 @@
+    return getResultUncached(request);
+  }
+
+  /// Produce the result of the request without caching.
```
s/result of/result/

huonw commented on include/swift/Basic/TypeID.h:
```diff
@@ -0,0 +1,99 @@
+SWIFT_TYPEID(C, float, 12);
+SWIFT_TYPEID(C, double, 13);
+SWIFT_TYPEID(C, bool, 14);
+  SWIFT_TYPEID(C, decltype(nullptr), 15);
```
Weird indentation?

huonw commented on include/swift/Sema/Requests.h:
```diff
@@ -0,0 +1,121 @@
+
+struct TypeLoc;
+
+/// Request the type from the ith entry in the inheritance clause for the
```
"index-th"

huonw commented on include/swift/Sema/Requests.h:
```diff
@@ -0,0 +1,121 @@
+  }
+};
+
+SWIFT_TYPEID(TypeChecker, InheritedTypeRequest, 1);
```
Doing these individually by hand seems prone to mistakes and problems with merge conflicts? Is it worth having (yet another) .def file?

DougGregor replied:
Yeah, I will do something here. It's an annoying pile of preprocessor metaprogramming.

huonw commented on lib/Basic/AnyValue.cpp:
```diff
@@ -0,0 +1,20 @@
+#include "swift/Basic/AnyValue.h"
+using namespace swift;
+
+AnyValue::HolderBase::~HolderBase() { }
```
This is the second definition of this: it's also above in lib/AST/Evaluator.cpp https://github.com/apple/swift/pull/15917/files#diff-dfb37be769b2d50ad4c150591bf84c59R22 ?

DougGregor replied:
That one is `AnyRequest::HolderBase::~HolderBase()`.

graydon commented on include/swift/AST/Evaluator.h:
```diff
@@ -0,0 +1,264 @@
+  typename Request::OutputType getResultUncached(const Request &request) {
+    // Push this request onto the active-requests stack while we
+    // evaluate it.
+    activeRequests.push_back(request);
```
Why push/pop it if you can't / aren't going to detect cycles involving it?

DougGregor replied:
I'm diagnosing cycles based on the stack, but you're right (per above) that I should use this for cycle detection as well.

graydon commented on include/swift/Basic/AnyValue.h:
```diff
@@ -0,0 +1,118 @@
+
+namespace swift {
+
+/// Stores a value of any type satisfies a small set of requirements
```
s/any type satisfies/any type that satisfies/

graydon commented on include/swift/Basic/TypeID.h:
```diff
@@ -0,0 +1,99 @@
+
+/// Form a type ID given a zone and type value.
+constexpr uint8_t formTypeID(uint8_t zone, uint8_t type) {
+  return (zone << 8) | type;
```
I'm probably doing my own arithmetic wrong, here but isn't a uint8_t shifted left by 8 bits always going to be 0? Maybe this should be returning a uint64_t?

DougGregor replied:
This should absolutely be returning a `uint64_t`!

graydon commented on include/swift/Basic/TypeID.h:
```diff
@@ -0,0 +1,99 @@
+                                                                        \
+public:                                                                 \
+  static const uint64_t value =                                         \
+    (templateID << (depth * 8)) | TypeID<Arg1>::value;                  \
```
Again here, I _think_ the arithmetic is off, and you want to shift by (depth * 16). Assuming you mean to have [256 zones | 256 types] packed as a 16 bits in each of up-to-4 depths, in a 64-bit word.

DougGregor replied:
Funny how it's the simple bit-manipulation math that I completely screwed up in this pull request. Thank you (and @huonw)

graydon reviewed (commented):
Bunch of nits but looks good!

huonw commented on include/swift/AST/Evaluator.h:
```diff
@@ -0,0 +1,264 @@
+///       When true, the request itself must provide an way to cache the
+///       results, e.g., in some external data structure. External caching
+///       should only be used when staging in the use of the evaluator into
+///       existing mutable data structures; new computations should not depend
```
Hm! Are we going to migrate to, say, every `TypeLoc` going through the `Evaluator`? If so, it kinda seems like we may want something with a more continuous memory layout? I'm thinking of basically doing an AoS-to-SoA transform, something like:

```cpp
class Evaluator {
    // ...

    llvm::DenseMap<TypeId, AnyCache> caches;
    // ...
}

class AnyCache {
    // a type erased `llvm::DenseMap<Request, Request::OutputType>`
}
```

or, even, going fully-static/killing our compile times:

```cpp
template<typename Request...>
class Evaluator {
    std::tuple<llvm::DenseMap<Request, Request::OutputType>...> caches;
    // or whatever the template magic to get a DenseMap for each request type is.
    // (plus extra magic to elide the map for externally cached requests.)
}

using SwiftEvaluator = Evaluator< InheritedTypeRequest, SuperclassTypeRequest, ...>;
```

The `getResultCached` function would then first need to find the appropriate cache (a downside of the first approach is, of course, that this may be moderately expensive and so outweighs the benefits of that style vs. the `[AnyRequest : AnyValue?]` version), but I think the rest is the same.

DougGregor replied:
This is a fantastic idea! Not the fully-static one, which would kill our layering and our compile times, but the type-erased `AnyCache` would be much better.

DougGregor committed:
Address the simple issues raised by Graydon and Huon

DougGregor committed:
[Evaluator] Detect cycles via the active-requests stack rather than the cache.

Turn the `activeRequests` stack into a `SetVector` and use it to detect
cycles. This approach works for uncached requests, and lets us simplify
the external-caching protocol quite a bit because we no longer record
“in-flight” states. Simplify the cache as well.

Thanks to Graydon for the suggestion.

DougGregor commented:
Some feedback addressed, but I haven't simplified `TypeID` definitions yet nor have I implemented @huonw 's `AnyCache` suggestion.

DougGregor committed:
[TypeID] Use .def files to introduce TypeID specializations.

Simplify the static registration of types for use with TypeID by introducing
a more declarative approach. Each zone provides a .def file listing the
types and templates defined by that zone. The .def file is processed by
include/swift/Basic/DefineTypeIDZone.h with its zone number, which assigns
values to each of the types/templates and introduces the TypeID
specializations.

DougGregor committed:
[Evaluator] Simplify the definition of new requests and make them printable.

Introduce a CRTP base class, SimpleRequest, which simplifies the task of
defining a new request kind by handling the storage of the values (in a
std::tuple), their hashing, equality, printing, etc. The values are passed
along to the subclass’s operator() so they’re mostly treated as (const)
parameters, making mutation of the request state impossible.

Extend AnyValue and AnyRequest with printing logic, so we can print any
request for debugging purposes, and

huonw commented:
The double layer of X-macros for the zones is... slightly mind-bending, but makes sense.

rudkx commented on include/swift/Basic/DefineTypeIDZone.h:
```diff
@@ -0,0 +1,84 @@
+//
+//    SWIFT_TYPEID_ZONE: The name of the Zone being defined, e.g.,
+//    SwiftAST for the Swift AST library's zone. All zones need to be defined
+//    a priori as 
```
....as....?

DougGregor replied:
Fixed with the latest update, thanks

DougGregor committed:
[Standard library] Remove incorrect FIXME.

DougGregor committed:
[Evaluator] Track dependencies of each request.

Perform online tracking of the (direct) dependencies of each request made
to the evaluator, and introduce debugging code to print these dependencies
as a tree (from a particular request).

DougGregor committed:
[Evaluator] GraphViz printing for all of the dependencies.

Introduce another form of debugging dump for the evaluator, rendering the
complete dependency graph using GraphViz, including all dependencies and
values cached within the evaluator.

DougGregor committed:
[Evaluator] Print cached value as part of dependency tree printing.

DougGregor committed:
[TypeID Fix comment regarding SWIFT_TYPEID_ZONE

DougGregor commented:
Several goodies in the latest update:

* All requests (and all values) are printable, so we can inspect anything

* Added `SimpleRequest`, which takes care of defining a new request with a simple payload (list of types) and automates basically all of it.

* Online dependency tracking

* Printing of the dependencies for a given request (as a tree), along with cached results, e.g.,
```
 `--InternallyCachedEvaluationRule(Binary: product) -> 2.461145e+02
     `--InternallyCachedEvaluationRule(Binary: sum) -> 5.859870e+00
     |   `--InternallyCachedEvaluationRule(Literal: 3.141590e+00)
     |   `--InternallyCachedEvaluationRule(Literal: 2.718280e+00)
     `--InternallyCachedEvaluationRule(Literal: 4.200000e+01)
```

* Printing all of the dependencies known to the evaluator as a GraphViz graph, along with known cached results, e.g.,

```
digraph Dependencies {
  request_0 -> request_1;
  request_0 -> request_4;
  request_1 -> request_3;
  request_1 -> request_2;
  request_5 -> request_6;
  request_5 -> request_9;
  request_6 -> request_8;
  request_6 -> request_7;
  request_10 -> request_11;
  request_10 -> request_14;
  request_11 -> request_13;
  request_11 -> request_12;

  request_0 [label="ExternallyCachedEvaluationRule(Binary: product)"];
  request_1 [label="ExternallyCachedEvaluationRule(Binary: sum)"];
  request_2 [label="ExternallyCachedEvaluationRule(Literal: 2.718280e+00)"];
  request_3 [label="ExternallyCachedEvaluationRule(Literal: 3.141590e+00)"];
  request_4 [label="ExternallyCachedEvaluationRule(Literal: 4.200000e+01)"];
  request_5 [label="InternallyCachedEvaluationRule(Binary: product) -> 2.461145e+02"];
  request_6 [label="InternallyCachedEvaluationRule(Binary: sum) -> 5.859870e+00"];
  request_7 [label="InternallyCachedEvaluationRule(Literal: 2.718280e+00)"];
  request_8 [label="InternallyCachedEvaluationRule(Literal: 3.141590e+00)"];
  request_9 [label="InternallyCachedEvaluationRule(Literal: 4.200000e+01)"];
  request_10 [label="UncachedEvaluationRule(Binary: product)"];
  request_11 [label="UncachedEvaluationRule(Binary: sum)"];
  request_12 [label="UncachedEvaluationRule(Literal: 2.718280e+00)"];
  request_13 [label="UncachedEvaluationRule(Literal: 3.141590e+00)"];
  request_14 [label="UncachedEvaluationRule(Literal: 4.200000e+01)"];
}
```

DougGregor committed:
[ChangeLog] Add an entry for SE-0194: Derived Collection of Enum Cases

DougGregor committed:
[Evaluator] Simplify the definition of cycle diagnostics for a SimpleRequest.

The arithmetic evaluator cannot really make use of the new diagnostics
machinery, but put it in place for other request kinds.

DougGregor committed:
[Evaluator] Add a Caching template parameter to SimpleRequest.

Simplify the definition of new request kinds by pulling the parameterization
of caching logic into SimpleRequest itself, so that the caching-related
parts of the contract cannot easily be forgotten.

DougGregor commented:
@swift-ci please smoke test

DougGregor committed:
[Evaluator test] Pull in std::isnan.

DougGregor commented:
@swift-ci please smoke test and merge

DougGregor commented:
@swift-ci please smoke test and merge
